collector/src/bin/rustc-perf-collector/execute.rs

```diff
@@ -500,6 +503,66 @@ impl<'a> Processor for ProfileProcessor<'a> {
                 panic!("unexpected profiler");
             }
 
+            // -Zself-profile produces (via rustc-fake) a data directory called
+            // 'Zsp' containing three files with names of the form
+            // `$BENCHMARK-$PID.{events,string_data,string_index}`. We copy it
+            // from the temp dir to the output dir, renaming the files within
+            // as `Zsp.{events,string_data,string_index}` in the process, then
+            // post-process them with `summarize`, `flamegraph`, and `crox` to
+            // produce several data files in the output dir.
+            Profiler::SelfProfile => {
+                let tmp_zsp_dir = filepath(data.cwd.as_ref(), "Zsp");
+                let zsp_dir = filepath(self.output_dir, &out_file("Zsp"));
+                let zsp_files_prefix = filepath(&zsp_dir, "Zsp");
+                let summarize_file = filepath(self.output_dir, &out_file("summarize"));
+                let flamegraph_file = filepath(self.output_dir, &out_file("flamegraph"));
+                let crox_file = filepath(self.output_dir, &out_file("crox"));
+
+                // Move the directory.
+                if zsp_dir.exists() {
+                    fs::remove_dir_all(&zsp_dir)?;
+                }
+                fs::rename(&tmp_zsp_dir, &zsp_dir)?;
+
+                // Rename the data files.
+                for entry in fs::read_dir(&zsp_dir).unwrap() {
+                    let filename = entry.unwrap().file_name();
+                    let filename_str = filename.to_str().unwrap();
+                    let path = filepath(&zsp_dir, filename_str);
+                    if filename_str.ends_with(".events") {
+                        fs::rename(path, filepath(&zsp_dir, "Zsp.events"))?;
+                    } else if filename_str.ends_with(".string_data") {
+                        fs::rename(path, filepath(&zsp_dir, "Zsp.string_data"))?;
+                    } else if filename_str.ends_with(".string_index") {
+                        fs::rename(path, filepath(&zsp_dir, "Zsp.string_index"))?;
```

Comment:
We should probably try to make sure we're not running over pre-existing files here (if we ran with -Zself-profile more than once)

Reply:
This is consistent with how things work for all the existing profilers -- if you specify the same parameters, it will overwrite things.

Reply:
Well, my concern was that we clobber ourselves (i.e. two rustc processes in a single benchmark get fun with -Zself-profile). Hopefully that doesn't actually happen, I guess.

Reply:
Oh, I see. I'm having trouble imagining the exact scenario that could cause clobbering like that, can you elaborate? The benchmark harness uses `cargo rustc -- -Zself-profile` so only the final crate is run with `-Zself-profile`...

Thanks!

Reply:
(When I was partway through writing the code I ran it a few times without doing the file renaming and I got just one set of three files per run. But I was using `helloworld` as my test program, so I may have missed some complication that could arise in more complex benchmarks.)

Reply:
I think it might be that I was putting everything in one directory, or so, which led to problems. I would feel better with an assert here I think, just to make sure we're not getting it wrong. But if we don't that's fine too, I agree we shouldn't have problems with cargo rustc in particular.

Reply:
See #621.